core/src/jsMain/kotlin/Advertisement.kt

```diff
@@ -1,12 +1,50 @@
 package com.juul.kable
 
+import com.benasher44.uuid.Uuid
+import com.juul.kable.external.BluetoothAdvertisingEvent
 import com.juul.kable.external.BluetoothDevice
+import com.juul.kable.external.JsIterator
+import com.juul.kable.external.iterable
+import org.khronos.webgl.DataView
 
 public actual class Advertisement internal constructor(
-    internal val bluetoothDevice: BluetoothDevice,
-    public actual val rssi: Int,
+    private val advertisement: BluetoothAdvertisingEvent,
 ) {
+    internal val bluetoothDevice: BluetoothDevice
+        get() = advertisement.device
 
     public actual val name: String?
         get() = bluetoothDevice.name
+
+    public actual val rssi: Int
+        get() = advertisement.rssi ?: 0
+
+    public actual val txPower: Int?
+        get() = advertisement.txPower
+
+    public actual val uuids: List<Uuid>
+        get() = advertisement.uuids.map { it.toUuid() }
+
+    public actual fun serviceData(uuid: Uuid): ByteArray? =
+        serviceDataAsDataView(uuid)?.buffer?.toByteArray()
+
+    public actual fun manufacturerData(companyIdentifierCode: Short): ByteArray? =
+        manufacturerDataAsDataView(companyIdentifierCode)?.buffer?.toByteArray()
+
+    public fun serviceDataAsDataView(uuid: Uuid): DataView? =
+        advertisement.serviceData.asDynamic().get(uuid.toString()) as? DataView
+
+    public fun manufacturerDataAsDataView(companyIdentifierCode: Short): DataView? =
+        advertisement.manufacturerData.asDynamic().get(companyIdentifierCode.toString()) as? DataView
+
+    public actual val manufacturerData: ManufacturerData?
+        get() = manufacturerDataEntries(advertisement.manufacturerData).firstOrNull()?.let { entry ->
+            ManufacturerData(
+                entry[0] as Short,
+                (entry[1] as DataView).buffer.toByteArray()
+            )
+        }
 }
+
+private fun manufacturerDataEntries(jsObject: dynamic): Iterable<Array<Any?>> =
+    (jsObject.entries() as JsIterator<Array<Any?>>).iterable()
```

Comment:
We can remove the "unsafe" cast if we define the structure of the JS advertisement (in `BluetoothAdvertisingEvent.kt`):

```kotlin
internal interface BluetoothManufacturerDataMap {
    fun entries(): JsIterator<Array<Any?>>
}

/**
 * https://webbluetoothcg.github.io/web-bluetooth/#bluetoothadvertisingevent
 */
internal abstract external class BluetoothAdvertisingEvent : Event {
    val device: BluetoothDevice
    val uuids: Array<String>
    val name: String?
    val rssi: Int?
    val txPower: Int?
    val manufacturerData: BluetoothManufacturerDataMap
    val serviceData: Any?
}
```

```suggestion
    public actual val manufacturerData: ManufacturerData?
        get() = advertisement.manufacturerData.entries().iterable().firstOrNull()?.let { entry ->
            ManufacturerData(
                entry[0] as Short,
                (entry[1] as DataView).buffer.toByteArray()
            )
        }
}
```

Maybe we could also define the types of the entries so we don't have to do the casting when we create our `ManufacturerData`).